### DISCUSSION TITLE
Button drawing "bug" for "no box" and "frame" buttons

### DISCUSSION THREAD

**Discussion by dannye** [fltk/fltk#1130] [Q&A]
This is probably the most minor thing I have reported so far, but I have always wondered so I wanted to ask about it.

Any button that has "no box" or a "frame" box instead of a "full" box will have its label redrawn without the "background" being redrawn first if the mouse is pressed and held on the button and then the mouse cursor is moved outside the button, resulting in "bold" looking text.

The label will also be "stuck" bold once the mouse button is released until the next full redraw.

For example, this can be seen with ordinary check buttons and radio buttons:

https://github.com/user-attachments/assets/87755ecc-9c3c-4e3f-b2bb-2abe0f8d985d

Also, any button that has "no box" but _does_ have a down_box will be "stuck" with its down_box if the mouse button is released while the mouse cursor is outside the button. (I have a few buttons like this in my application.)

Are these known issues?

**Comment by MatthiasWM** [comment]
Yes, this is a known effect of not drawing a background. The antialiased font is drawn multiple times at the same position, causing thickened out edges. Having no box or a frame-only box does not make sense for most widgets. The none or frame types can be used for widgets that draw their own background somehow (custom code, filled with children).

**Comment by dannye** [comment]
Hm, I can understand the reasoning, but these aren't really rare/special cases given how ordinary check buttons and radio buttons are.
So solving this would still be desirable, right?

And in the case of my application, I have a few "subtle" buttons that use no box, but a regular/full down_box that are susceptible to getting "stuck" bold and drawn with their down_box because of this bug. These buttons can't draw their own background because they show the texture of their parent.

https://github.com/user-attachments/assets/5f62497e-813e-4d0a-8c51-f144797ac206

**Comment by Albrecht-S** [comment]
> solving this would still be desirable, right?

I hope we can solve this in 1.5.0. My vision is to set a boxtype attribute like "has background" (bool) or one of {`background`|`frame`|`no box`}. When drawing a label we might "know" that this is either in a `no-background` box or outside the widget's box. In both cases we'd have to draw the parent's background - and this recursively until we find an enclosing widget with a real background (box). Something like that...

**Comment by MatthiasWM** [comment]
Oh, I see now. Yeah, try to replace line 125 in src/Fl_Button.cxx (case FL_DRAG in Fl_Button::handle()):
```
//        redraw();
        if (box()) redraw();
        else redraw_label();
```
and let me know if this fixes it for all cases. It probably won't for frame types, so maybe this is better:

```
        if (box() && (fl_box(box())==box())) redraw();
        else redraw_label();
```
If it is, it needs to be changed in Fl_Button::value() as well.

**Reply by dannye** [reply]
Both of those seem like a fantastic improvement to me, for both test/radio and my custom buttons.

**Comment by MatthiasWM** [comment]
Thanks for the confirmation. This is just a hot fix. I'll have to check if there is a better and complete fix out there. I should be able to get something in in the next three days,

**Reply by Albrecht-S** [reply]
@MatthiasWM Ping ;-)
The "next three days" have become three weeks. I don't want to push, but do you think that it's reasonable to pursue this for 1.4.x (x > 1)? I don't want to hold release 1.4.1 for sure, but can we consider an improvement in the near future?

The full fix (drawing the bg of the parent(s) recursively) should be reserved to 1.5 though. IMHO.

See also new issue #1165 (this is a FAQ).

**Reply by MatthiasWM** [reply]
I wasted hours on this, but  in the end, it always comes out to "use a box type, not a framtype". I will apply the suggested patch as it seems to satisfy the default cases. I absolutely do not want to hold up 1.4.1! Just slipped off the radar.

**Comment by MatthiasWM** [comment]
Ok, so I patched this with the drawing code I suggested earlier. This should work well.

**Reply by Albrecht-S** [reply]
@MatthiasWM Thanks. I updated #1165 and assigned it to you.
